Source/JavaScriptCore/wasm/WasmOperationsInlines.h

```diff
@@ -192,6 +195,59 @@ inline void structSet(Instance* instance, EncodedJSValue encodedStructReference,
     return structPointer->set(globalObject, fieldIndex, toJSValue(globalObject, fieldType, argument));
 }
 
+inline bool refCast(Instance* instance, EncodedJSValue encodedReference, bool allowNull, int32_t heapType)
+{
+    JSValue refValue = JSValue::decode(encodedReference);
+    if (refValue.isNull())
+        return allowNull;
+
+    switch (static_cast<Wasm::TypeKind>(heapType)) {
+    case Wasm::TypeKind::Funcref:
+    case Wasm::TypeKind::Externref:
+        // Casts to funcref/externref canot fail as they are the top types of their respective hierarchies, and static type-checking does not allow cross-hierarchy casts.
+        return true;
+    case Wasm::TypeKind::I31ref:
+        return refValue.isInt32();
+    case Wasm::TypeKind::Arrayref:
+        return jsDynamicCast<JSWebAssemblyArray*>(refValue);
+    case Wasm::TypeKind::Structref:
+        return jsDynamicCast<JSWebAssemblyStruct*>(refValue);
+    default: {
+        ASSERT(!Wasm::typeIndexIsType(static_cast<Wasm::TypeIndex>(heapType)));
+        Wasm::TypeDefinition& signature = instance->module().moduleInformation().typeSignatures[heapType];
+        auto signatureRTT = instance->module().moduleInformation().rtts[heapType];
+        if (signature.expand().is<Wasm::FunctionSignature>()) {
+            WebAssemblyFunctionBase* funcRef = jsDynamicCast<WebAssemblyFunctionBase*>(refValue);
+            // Static type-checking should ensure this jsDynamicCast always succeeds.
+            ASSERT(funcRef);
+            auto funcRTT = funcRef->rtt();
+            if (funcRTT.get() == signatureRTT.get())
+                return true;
+            return funcRTT->isSubRTT(*signatureRTT);
+        }
+        if (signature.expand().is<Wasm::ArrayType>()) {
+            JSWebAssemblyArray* arrayRef = jsDynamicCast<JSWebAssemblyArray*>(refValue);
+            if (!arrayRef)
+                return false;
+            auto arrayRTT = arrayRef->rtt();
```

Comment:
For structs and arrays this just accesses the RTT from the object directly (before it looked up the Structure, then RTT in the structure)